README.md
# RMH Covid-19 Home Monitoring Project
RMH Home Monitoring lets a hospital manage a large number of Covid-19 patients using their smartphones and two simple medical devices (pulse oximeter + thermometer).

The goal is to oversee a large number of relatively healthy patients in their own homes, saving bed space and staff resources.

The system prompts patients twice per day to submit their own heart rate, body temperature and oxygen saturation via a website. Clinicians are alerted to patients with concerning or critical vitals. Automation keep staff interaction and administrative work minimal.

RMH Home Monitoring is a [REDCap](https://projectredcap.org/software/) (Research Electronic Data Capture tool) project template that you can copy and adapted for your facility. Patients interact with the system via public websites (REDCap *Surveys*).

## Features

- Patient registration
- Automatically prompt patients for their heart rate, temperature and oxygen saturation
- Store and analyse these 'Observations'
- Alert staff to patients that should be reviewed because of high/low vitals
- Alert staff if a patient should return to the hospital immediately (a MET call)
- Alert patient and staff of patients who are late submitting their data
- Alert thresholds are set per patient
- Keep clinical notes for each patient
- Discharge patients and produce medical records
- Ask various other tracking questions for medical analysis

![Screenshot](https://user-images.githubusercontent.com/1585749/81142763-d1cccf80-8fb3-11ea-9c00-5e8d08e0de3d.png)


**As of April 2020** the project is in use at the Royal Melbourne Hospital in Victoria, Australia.

This project has been developed by Dr Martin Dutch and Associate Professor Jonathan Knott and is shared with compliments of [The Royal Melbourne Hospital, Victoria, Australia](https://www.thermh.org.au/). [The Peter Doherty Institute for Infection and Immunity](https://www.doherty.edu.au/) has provided funding for the project through a research grant.

You may use and adapt this project free of charge, however we ask that you acknowledge this project and the Royal Melbourne Hospital in your copy, and any associated academic publications.

We would be delighted to received correspondence if your health service has found our contribution useful, and would be happy to work collaboratively to measure the impact of this service delivery. Contact details are below.

Though a non-traditional tool for building a monitoring system, REDCap allowed us to rapidly prototype the system in response to the pandemic in early 2020. We hope it will also make the system accessible to organisations facing Covid-19 pressure but without the resources to build software projects.

## Requirements for using the system

*You must have your own running REDCap system to use this tool. Unfortunately we are not in a position to host your institution's health data.*

- a [REDCap](https://projectredcap.org/software/) installation and license
- a recent enough version of REDCap to support the *Alerts and Notifications Function*
- a mobile phone that can receive SMS for staff alerts
- an outgoing SMS service (ideally Twilio)
- an outgoing email server and an email address for your staff
- ability to provide patients with a digital thermometer and pulse oximeter
- staff to respond to patient observations and alerts
- staff to manage patient registrations and discharge

## Download, setup the tool, and prepare your team

Take a copy of our REDCap project file. **You must modified it** before you use it.

1. Read the brief [Overview Guide](https://github.com/rmhcovid/txtmon/blob/master/documentation/overview_START_HERE.md).
2. Read the [Home Monitoring Program Guide](https://github.com/rmhcovid/txtmon/blob/master/latest_version/RMH%20Covid-19%20Home%20Monitoring%20Program.pdf) that explains how the project is run, the staff and patient processes.
3. Follow the [Setup Instructions](https://github.com/rmhcovid/txtmon/blob/master/documentation/setup_instructions.md) to get a copy up and running.

## Help and Known Issues

Please see a list of known issues in the [Setup Instructions](https://github.com/rmhcovid/txtmon/blob/master/documentation/setup_instructions.md).

## How to Contribute

We'd love to get your feedback and to hear about how you've been able to use this work in your own organisations. You can email us at rmhcovid@314.net.au.

## Reporting Issues

We are not in a position to offer technical support. We are offering our work so far in the hope that it might be useful for other organisations. You may run something similar to what we have, or adapt it for your needs.

You are welcome to add Issues using the normal GitHub process, however you *should not expect to get a reply*.

We will consider any formal pull requests made via the usual GitHub process.


## Legal

Please see the attached LICENSE document.

## Version History

- See  [CHANGELOG.md](https://github.com/rmhcovid/txtmon/blob/master/CHANGELOG.md)
- 1.1.0 - 2020-05-20 - COVIDHomeMonitoring_2020-05-20_1527.REDCap.xml
- 0.9.0 - 2020-05-05 - CovidHomeMonitoring_2020-05-05_1649.REDCap.xml

## Author and Acknowledgements

This project was designed and managed using the [REDCap](https://projectredcap.org/software/) electronic data capture tool hosted by the [Royal Melbourne Hospital Business Intelligence Unit](https://www.thermh.org.au/). [The Peter Doherty Institute for Infection and Immunity](https://www.doherty.edu.au/) has provided funding through a research grant to make this project possible.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.


## [1.1.0] - 2020-05-20

### Added
- Streamlined sign-up process
- A patient registration and consent process
- Support for 'Ineligible' patients (recorded, but not monitored)
- Registration confirmation page
- Discharge Required now links staff to discharge form
- Various audit tests
- Tool to authenticate redcap session

### Changed
- Re-ordered many fields in Registration instrument
- Patient Age now based on [registration_date] not [admission_date]
- Pre-existing Condition fields changed from checkboxes to radiobuttos
- Daily observation prompts logic rebuilt
- Separated discharge into separate instrument / survey
- Discharge Required trigger logic changed
- Monitoring status now based on monitoring group and discharge form
- Field [mon_discharge_reason] now defaults to 'HEALTHY'
- Standardised naming of COVID (not Covid) across the project
- Rebuilt alert_update tool to remove need for curl-copy-pastes

### Deprecated - things we intend to take away
- WEEKLY monitoring group patients

### Fixed
- Patient can now skip an observation without getting stuck


## [0.9.0] - 2020-05-05

### Added
- First public release of the system


[1.1.0]: https://github.com/rmhcovid/txtmon/compare/v0.9.0...v1.1.0
[0.9.0]: https://github.com/rmhcovid/txtmon/releases/tag/v0.9.0


tools/authenticate.py
#!/usr/bin/env python
"""
Authenticate with your REDCap website to get a PHPSESSID and redcap_csrf_token
"""

import re
import sys

import requests


def attempt_login(url, username, password):
    """
    Attempt to login to the given REDCap server admin site and retrieve
    a valid PHP session id and CSRF token.
    """

    # Login requires a token included in the html form
    # Kind of like a csrf token (but different to the real csrf used elsewhere)
    res = requests.get(url)
    login_token = re.search(r"id='(redcap_login_[^']+)'", res.text).group(1)

    # Data required for authentication
    data = {
        'username': username,
        'password': password,
        'submitted': 1,
        login_token: '',
    }

    # Use a Requests session object so we can capture the cookies
    sess = requests.Session()
    res = sess.post(url, data=data)

    # REDcap doesn't give any http header indication of a failed login
    # only the page text changes.
    if 'redcap_csrf_token' in res.text:
        session_cookie = sess.cookies['PHPSESSID']
        csrf_token = re.findall(r"var redcap_csrf_token = '([^']+)'", res.text)[0]
    else:
        session_cookie = None
        csrf_token = None

    return (session_cookie, csrf_token)


if __name__ == '__main__':

    if len(sys.argv) != 4:
        sys.exit("Usage:\n%s URL USERNAME PASSWORD" % sys.argv[0])

    url = sys.argv[1]
    username = sys.argv[2]
    password = sys.argv[3]
    (session, csrf) = attempt_login(url=sys.argv[1],
                                    username=sys.argv[2],
                                    password=sys.argv[3])

    if not (session and csrf):
        sys.exit("Could not authenticate")

    sys.stdout.write(f"SESSIONID='{session}'\n")
    sys.stdout.write(f"CSRF_TOKEN='{csrf}'\n")


tools/generate_alerts/list_alerts.py
#! /usr/bin/env python
"""
List all the Alerts defined in a REDCap

This tool should output something like this:
  152 1 SMS Patient Registration - BIDAILY
  165 2 SMS Patient Registration - ESCALATION checkin
  175 3 SMS Late observation patient emergency contact - SMS
  176 4 EMAIL Late observation staff warning - EMAIL
  177 5 SMS Monitoring Discharge patient - SMS
  178 6 EMAIL Monitoring Discharge patient - EMAIL
  179 7 EMAIL Monitoring Discharge Required staff
  180 8 SMS Escalation checkin warning patient
  181 9 SMS Periodic info note - Public Health Messaging SMS
  182 10 SMS Periodic info note - Quarantine SMS
  183 11 SMS Periodic info note - Lifeline SMS
  184 12 SMS Periodic info note - Meals on Wheels
  200 13 SMS Obs Combined Staff alert - Ob_template
  201 14 SMS Obs Combined Patient alert - Ob_template
  202 15 SMS Obs Combined Patient alert - Ob_0
  203 16 SMS Obs Combined Patient alert - Ob_1a
  ...


Usage
  ./list_alerts.py PROJECTID PHPSESSIONID
  ./list_alerts.py 984 6437fdsjadfskh2fdl23k4j323


It works by issuing a GET request for the Alerts & Notifications page in
REDCap, and parsing the html for the alerts and metadata.
"""

from urllib.parse import parse_qs
import json
import re
import sys

import requests

URL = "https://redcap.yourcompay.com/redcap_v9.8.0/index.php"


def list_alerts(url, phpsessionid, projectid):
    """Load the REDCap Alerts & Notifications tab, parse out the list of alerts
    and their metadata and print it to stdout"""

    qwargs_template = "pid=nnn&route=AlertsController:setup"

    qwargs_data = parse_qs(qwargs_template)
    qwargs_data['pid'] = projectid

    headers = {
        'cookie': 'PHPSESSID=%s' % phpessionid,
    }

    res = requests.get(url, headers=headers, params=qwargs_data)
    # print(res)

    # Easiest place to find the title/id/index is in the javascript that
    # triggers the edit box for each alert
    # look for  function __rcfunc_editEmailAlert_emailRow....
    reg = r'__rcfunc_editEmailAlert_emailRow.*editEmailAlert\(({.*}),'

    matches = re.findall(reg, res.text)

    for match in matches:
        alert_data = json.loads(match)
        print(
            alert_data['alert-id'],
            alert_data['alert-number'],
            alert_data['alert-type'],
            alert_data['alert-title'],
        )


if __name__ == '__main__':

    if len(sys.argv) != 3:
        sys.exit("Usage:\n%s PROJECTID PHPSESSIONID" % sys.argv[0])

    projectid = sys.argv[1]
    phpessionid = sys.argv[2]

    list_alerts(url=URL, phpsessionid=phpessionid, projectid=projectid)


tools/audit_project/download_project_xml.py
#! /usr/bin/env python
"""
Connect to redcap and download the project metadata xml file
eg CovidHomeMonitoring_2020-04-16_1120.REDCap.xml

Downloads a metadata export of your project.
Includes Automated Survey Invites and leaves them enabled.

Equivalent to this in the REDCap admin:
- open project
- Project Home
- Other Functionality
- Copy or Backup Project
- Include the following in the XML file: (tick all)
- Download metadata only (XML)


Equivalent to this curl statement, captured from chrome dev tools.

curl 'https://redcapserver.yourcompany.org/redcap_v9.8.0/ProjectSetup/export_project_odm.php?pid=984&xml_metadata_options=userroles,reports,alerts,surveys,sq,asi,asienable' -H 'authority: redcap.yourcompay.com' -H 'upgrade-insecure-requests: 1' -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.163 Safari/537.36' -H 'sec-fetch-dest: document' -H 'accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9' -H 'sec-fetch-site: same-origin' -H 'sec-fetch-mode: navigate' -H 'sec-fetch-user: ?1' -H 'referer: https://redcap.yourcompay.com/redcap_v9.8.0/ProjectSetup/other_functionality.php?pid=984' -H 'accept-language: en-GB,en;q=0.9,en-US;q=0.8,la;q=0.7' -H 'cookie: survey=fdsljsfdljsdfjldfljfdjjdk6; PHPSESSID=aabbcacaafdaaaaadfafaaaaaa' -H 'dnt: 1' --compressed

"""

from pathlib import Path
from urllib.parse import parse_qs
import os
import re
import sys

import requests
from pathvalidate import sanitize_filename


EXPORT_URL = r"https://redcap.yourcompay.com/redcap_v9.8.0/ProjectSetup/export_project_odm.php"


def download_project_metadata(projectid, phpsessionid):

    querystring = "pid=984&xml_metadata_options=userroles,reports,alerts,surveys,sq,asi,asienable"

    # Sub in the correct project ID
    qwargs_data = parse_qs(querystring)
    qwargs_data['pid'] = projectid

    # Sub in our session ID to let our script authenticate
    headers = {
        'cache-control': 'max-age=0',
        'accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
        'cookie': 'PHPSESSID=%s' % phpessionid,
    }

    res = requests.get(url=EXPORT_URL, headers=headers, params=qwargs_data)
    # print(res)

    # Work out the filename it generated for us from the headers
    # eg 'Content-Disposition': 'attachment; filename="CovidHomeMonitoring_2020-04-16_1127.REDCap.xml"'
    filename = re.findall('filename="(.+)"', res.headers['content-disposition'])[0]

    # Sanitise the filename. Basic securtiy
    clean_filename = sanitize_filename(filename, replacement_text="_")
    newfile_path = Path(".") / clean_filename

    # Don't overwrite an existing file with the same name
    if os.path.exists(newfile_path):
        sys.exist(f"Cannot write xml file. File already exists: {clean_filename}")

    with open(newfile_path, 'wb') as f:
        f.write(res.content)

    sys.stdout.write(f"Project saved to {newfile_path}\n")


if __name__ == '__main__':

    if len(sys.argv) != 3:
        sys.exit("Usage:\n%s PROJECTID PHPSESSIONID" % sys.argv[0])

    projectid = sys.argv[1]
    phpessionid = sys.argv[2]

    # TODO validate projectid?
    # TODO phpsessionid
    download_project_metadata(projectid=projectid, phpsessionid=phpessionid)


tools/generate_observations/generate_obs_definitions.py
#! /usr/bin/env python

import sys


def generate_obs(template, unique_code):
    ob_name = "ob_%s" % unique_code
    field_follower = "_%s" % unique_code

    csv = template.replace("ob_template", ob_name)
    csv = csv.replace("_template", field_follower)

    return csv


if __name__ == '__main__':
    if len(sys.argv) != 2:
        sys.exit("Usage %s observation_template.csv" % sys.argv[0])

    obs_template = open(sys.argv[1]).read().strip()

    print(generate_obs(obs_template, "0"))
    print(generate_obs(obs_template, "1a"))
    print(generate_obs(obs_template, "1b"))
    print(generate_obs(obs_template, "2a"))
    print(generate_obs(obs_template, "2b"))
    print(generate_obs(obs_template, "3a"))
    print(generate_obs(obs_template, "3b"))
    print(generate_obs(obs_template, "4a"))
    print(generate_obs(obs_template, "4b"))
    print(generate_obs(obs_template, "5a"))
    print(generate_obs(obs_template, "5b"))
    print(generate_obs(obs_template, "6a"))
    print(generate_obs(obs_template, "6b"))
    print(generate_obs(obs_template, "7a"))
    print(generate_obs(obs_template, "7b"))
    print(generate_obs(obs_template, "8a"))
    print(generate_obs(obs_template, "8b"))
    print(generate_obs(obs_template, "9a"))
    print(generate_obs(obs_template, "9b"))
    print(generate_obs(obs_template, "10a"))
    print(generate_obs(obs_template, "10b"))
    print(generate_obs(obs_template, "11a"))
    print(generate_obs(obs_template, "11b"))
    print(generate_obs(obs_template, "12a"))
    print(generate_obs(obs_template, "12b"))
    print(generate_obs(obs_template, "13a"))
    print(generate_obs(obs_template, "13b"))
    print(generate_obs(obs_template, "14a"))
    print(generate_obs(obs_template, "14b"))


tools/generate_alerts/update_alert.py
#! /usr/bin/env python
"""
Connect to the REDCap admin site and update a specific existing alerts settings

Use this tool to update an alert based on another template alert. We use it
to copy out all the various ob_0, ob_1a, ob_1b, ... alerts.

THIS IS NOT A GENERIC TOOL. It relies on lots of developer input and update.
But it's better than clicking 400 buttons in the admin UI.

This injects http POST requests into the server. Authenticates by using an
existing PHP session cookie and a CSRF token. See the documentation for how
to get these.


Usage
  ./update_alert.py phpsession csrftoken project type template target obcode

ie update alert ALERTINDEX to match the template ALERT_TYPE, but substitute
the appropriate values for OB_COD.

eg

  ./update_alert.py d7kbb39c... 31ee8469.... 984 STAFF_ALERT_COMBINED_SMS 200 290 ob_14b
"""

import json
import sys
import re
from pprint import pprint
from urllib.parse import parse_qs

import requests


# These are the html post vars required to update an alert
# they map *almost* 1-to-1 with a structure set by javascript inside the
# alert page editEmailAlert({...}, ...)
POST_TEMPLATE = {
    # REDCap seems to pull these from the database
    # They are all present in the script call
    'alert-type',
    'email-to',
    'ensure-logic-still-true',
    'cron-send-email-on-date',
    'alert-expiration',
    'cron-send-email-on-next-time',
    'email-failed',
    'email-bcc',
    'email-repetitive-change-calcs',
    'cron-repeat-for',
    'email-deleted',
    'redcap_csrf_token',
    'phone-number-to',
    'alert-stop-type',
    'alert-title',
    'email-subject',
    'cron-send-email-on-time-lag-days',
    'email-from',
    'cron-send-email-on',
    'alert-message',
    'cron-repeat-for-units',
    'email-cc',
    'alert-condition',
    'cron-repeat-for-max',
    'cron-send-email-on-time-lag-minutes',
    'email-attachment-variable',
    'email-incomplete',
    'cron-send-email-on-next-day-type',
    'form-name',
    'email-repetitive',
    'email-repetitive-change',
    'cron-send-email-on-time-lag-hours',
    'email-from-display',

    # REDCap set by JS during the save
    'alert-send-how-many',
    'index_modal_update',
    'phone-number-to-freeform',
    'email-to-freeform',
    'email-cc-freeform',
    'alert-trigger',
    'email-bcc-freeform',

    # REDCap UI housekeeping
    'redcap_csrf_token',
    'alert-message-editor',
}


ALERT_TEMPLATE_ID = {
    'STAFF_ALERT_COMBINED_EMAIL': '216',
    'STAFF_ALERT_COMBINED_SMS': '200',
    'PATIENT_ALERT_COMBINED_SMS': '201',
    'LATE_OBS_STAFF_SMS': '307'
}


def update_alert(sess, template_postdata, alert_type, csrftoken, projectid, alert_index, obcode):

    # Different alerts need us to transform the template in different ways
    # Some are a simple string.replace('_template', '_4a'), some just be set
    # explicitly
    if alert_type in ['STAFF_ALERT_COMBINED_EMAIL', 'STAFF_ALERT_COMBINED_SMS',
                      'PATIENT_ALERT_COMBINED_SMS']:
        transformed_data = prepare_postvars_obs_simplealert(template_postdata, obcode)
    elif alert_type == 'LATE_OBS_STAFF_SMS':
        transformed_data = prepare_postvars_late_obs_staff(template_postdata, obcode)
    else:
        sys.exit(f"Unknown alert_type: {alert_type}")


    # Form update stuff
    transformed_data['index_modal_update'] = alert_index
    transformed_data['redcap_csrf_token'] = csrftoken

    for key, val in transformed_data.items():
        print(f"{key}: {val}")

    URL = "https://redcap.yourcompay.com/redcap_v9.8.0/index.php"
    QVAR_TEMPLATE = "pid=984&route=AlertsController:saveAlert"

    template_qvar_data = {k: v[0] for k, v in parse_qs(QVAR_TEMPLATE).items()}
    # TODO assert contains the values we expect?
    transformed_qvars_data = {k: v for k, v in template_qvar_data.items()}
    transformed_qvars_data['pid'] = projectid

    headers = {
        'content-type': 'application/x-www-form-urlencoded; charset=UTF-8',
    }

    payload = transformed_data
    pprint(payload)


    req = requests.Request(
        'POST', url=URL, headers=headers, params=transformed_qvars_data, data=payload)
    prep = req.prepare()
    prep.headers = {**sess.headers, **prep.headers}

    #breakpoint()
    res = sess.send(prep)

    checkexit_redcap_loggout(res)
    print(res)
    print(res.text)


def prepare_postvars_obs_simplealert(template_postdata, obcode):
    """In these alerts we just change anything _template to _n
    eg   [hr_template] -> [hr_4a]
    eg   "combined email ob_template" -> "combined email ob_4a"
    """

    obprefix, obcode_trailer = obcode.split('_')

    transformed_data = {}
    for key, vals in template_postdata.items():
        print(f"{key}: {vals}")

        tranformed_key = key.replace('_template', f'_{obcode_trailer}')

        # TODO what if that text exists as a literal in the text?
        # TODO do we test for undefined? [variables]?


        tranformed_vals = vals
        if tranformed_vals:
            tranformed_vals = vals.replace('_template', f'_{obcode_trailer}')

        transformed_data[tranformed_key] = tranformed_vals

    if transformed_data['phone-number-to-freeform'] is None:
        if transformed_data['phone-number-to'] > '':
            transformed_data['phone-number-to-freeform'] = transformed_data['phone-number-to']

            transformed_data['phone-number-to'] = None
    return transformed_data


def prepare_postvars_late_obs_staff(template_postdata, obcode):
    """Creat postdata for updating a late_obs_staff alert.
    This is trickier than the 'combined staff' alerts. They just need all
    the _template text changed to _OBSNUMBER

    But late obs alerts have references to multiple observations. They are
    triggered by one observation, and have the details of another.
    ie ob_3a triggers the alert to check ob_4a
    """
    preceding_obprefix, preceding_obnum = get_obs_preceding_obs(obcode).split("_")

    # Explicitly set the triggering form to be the PRECEEDING observation
    transformed_data = template_postdata
    transformed_data['form-name'] = transformed_data['form-name'].replace('_template', f'_{preceding_obnum}')

    # Now just replace the rest as usual
    transformed_data = prepare_postvars_obs_simplealert(transformed_data, obcode)

    # And set the alert time to be 5 hours after due time
    # 1pm for the morning, 8pm for the evening
    if obs_is_morning(obcode):
        transformed_data['cron-send-email-on-next-time'] = '13:00'
    elif obs_is_afternoon(obcode):
        transformed_data['cron-send-email-on-next-time'] = '20:00'
    else:
        assert False  # unknown obcode

    return transformed_data


# TODO code is doubled up with audit_project.py
def valid_obcode(code):
    # Special cases
    if code in ['ob_template', 'ob_0']:
        return True

    if '_' not in code:
        return False

    if not code[0:3] == 'ob_':
        return False

    # They shoudl all end in 'a' or 'b' excep ob_0 and ob_template
    if code[-1] not in ['a', 'b']:
        return False

    # Digits
    if not code[3:-1].isdigit():
        return False

    return True


# TODO code is doubled up with audit_project.py
def obs_is_morning(obcode):
    """Given an observation code (eg 'ob_1a', 'ob12_b') is this a morning obs?"""
    return obcode[-1] == 'a'


# TODO code is doubled up with audit_project.py
def obs_is_afternoon(obcode):
    """Given an observation code (eg 'ob_1a', 'ob12_b') is this an afternoon obs?"""
    return obcode[-1] == 'b'


# TODO code is doubled up with audit_project.py
def get_obs_preceding_obs(obcode):
    """Given an observation eg ob_4a, which observation precedes it?
    We have two streams, the 'a' (morning) and 'b' (afternoon) obs.

    But the first day follows on from the demo observation ob_0 -> ob_1a
                                                           ob_0 -> ob_1b
    The rest follow the previous. ob_1a -> ob_2a -> ob_3a
                                  ob_1b -> ob_2b -> ob_3b
    """
    assert valid_obcode(obcode)

    # First day follows the demo observation
    if obcode in ['ob_1a', 'ob_1b']:
        return 'ob_0'

    # The rest follow the previous day (number -1)
    ob_day = obcode[3:-1]
    prev_day = int(ob_day) - 1

    if obs_is_morning(obcode):
        return 'ob_%sa' % prev_day
    elif obs_is_afternoon(obcode):
        return 'ob_%sb' % prev_day
    else:
        assert False  # unknown obcode


def load_postdata_from_template(sess, projectid, template_id):
    alerts_page_url = get_alerts_page_url(projectid)
    res = sess.get(alerts_page_url, allow_redirects=False)
    checkexit_redcap_loggout(res)

    alerts_page_html = res.text
    data = extract_alert_data_structure(alerts_page_html, template_id)

    # Add the extra post fields

    # Update the triggering form to include event code
    # This is probably a bug. Not sure how this works with longitudinal
    data['form-name'] = "%s-%s" % (data['form-name'], data['form-name-event'])

    data = {
        'alert-send-how-many': None,
        'index_modal_update': template_id,
        'phone-number-to-freeform': None,
        'email-to-freeform': None,
        'email-cc-freeform': None,
        'alert-trigger': 'submit-logic',
        'email-bcc-freeform': None,
        'redcap_csrf_token': None,
        'alert-message-editor': None,
        **data
    }

    filtered = {k: data[k] for k in POST_TEMPLATE}

    assert set(POST_TEMPLATE) == set(filtered.keys())
    return filtered


def extract_alert_data_structure(alerts_page_html, alertid):
    """
    Find the javascript call to open the editor for alertid. Extract params.

    Looks something like this:
       <script type="text/javascript">function __rcfunc_editEmailAlert_emailRow10(){ editEmailAlert({"alert-id":"200","alert-title":"...},200,11) }</script>
    """
    for scriptcall in re.findall(r"editEmailAlert\({(.*)}.*\)", alerts_page_html):
        if f'"alert-id":"{alertid}"' in scriptcall:
            data = json.loads("{" + scriptcall + "}")
            return data

    raise Exception("Could not find template alert structure")


def get_alerts_page_url(projectid):
    # TODO break out to config
    return f"https://redcap.yourcompay.com/redcap_v9.8.0/index.php?pid={projectid}&route=AlertsController:setup"


def checkexit_redcap_loggout(response):
    """Given a requests response exit with error if we've been logged out.

    Redcap does not give a http response code on logout. You just get back
    Detect if we've been bumped to a login page and exit"""

    # Quick and dirty. Look for the forgot password link
    if response.status_code == 302 or 'Authentication/password_recovery.php' in response.text:
        sys.exit("Logged out of REDCap")


if __name__ == '__main__':

    if len(sys.argv) != 8:
        sys.exit(
            "Usage:\n%s phpsession csrftoken project type template target obcode" %
            sys.argv[0])

    phpsessionid = sys.argv[1]
    csrf_token = sys.argv[2]
    projectid = sys.argv[3]
    alert_type = sys.argv[4]
    template_id = sys.argv[5]
    target_id = sys.argv[6]
    obcode = sys.argv[7]

    # TODO validate phpessionid

    # TODO validate projectid

    # TODO validate csrf_token

    # Is this an alert type we know?
    if alert_type not in ALERT_TEMPLATE_ID.keys():
        sys.exit("Unknown alert type: %s\nKnown types: %s" % (alert_type, ALERT_TEMPLATE_ID.keys()))

    if not template_id.isdigit():
        sys.exit("Invalid template index: %s" % template_id)

    if not target_id.isdigit():
        sys.exit("Invalid target index: %s" % target_id)

    # TODO validate obcode
    if not valid_obcode(obcode):
        sys.exit("Invalid ob_code: %s" % obcode)


    # Authenticated requests session
    sess = requests.Session()
    sess.headers.update({'cookie': f'PHPSESSID={phpsessionid}'})



    template_postdata = load_postdata_from_template(sess, projectid, template_id)


    alert_title = template_postdata['alert-title']

    # Did we get the right template? Check the expected redcap object id
    # is the same one that's in the curl cmd template
    assert template_postdata['index_modal_update'] == ALERT_TEMPLATE_ID[alert_type]

    print("Does this look right?...")
    pprint({
        'phpsessionid': phpsessionid,
        'csrftoken': csrf_token,
        'projectid': projectid,
        'type': alert_type,
        'template_id': template_id,
        'template title': alert_title,
        'target_id': target_id,
        'obcode': obcode
    })

    # Make user confirm before actually sending the update
    if False:
        print("y/n...")
        confirm = input().lower()

        if confirm != 'y':
            sys.exit("Aborting")

    update_alert(
        sess,
        template_postdata=template_postdata,
        alert_type=alert_type,
        csrftoken=csrf_token,
        projectid=projectid,
        alert_index=target_id,
        obcode=obcode)
